tests/pricing_models/test_calc_in_given_out.py

```diff
@@ -0,0 +1,1097 @@
+#    6 mo remaining
+# 6. in_ = 200; 10% fee; 100k share reserves; 1M bond reserves;
+#    2 share price; 1.5 init share price; t_stretch targeting 5% APY;
+#    6 mo remaining
+# 7. in_ = 200; 10% fee; 100k share reserves; 1M bond reserves;
+#    2 share price; 1.5 init share price; t_stretch targeting 5% APY;
+#    3 mo remaining
+# 8. in_ = 200; 10% fee; 100k share reserves; 1M bond reserves;
+#    2 share price; 1.5 init share price; t_stretch targeting 10% APY;
+#    3 mo remaining
+base_in_test_cases = [
+    (  ## test one, basic starting point
+        TestCaseCalcInGivenOutSuccess(
+            out=Quantity(amount=100, unit=TokenType.PT),  # how many tokens you expect to get
+            market_state=MarketState(
+                share_reserves=100_000,  # base reserves (in share terms) base = share * share_price
+                bond_reserves=100_000,  # PT reserves
+                share_price=1,  # share price of the LP in the yield source
+                init_share_price=1,  # original share price pool started
+            ),
+            fee_percent=0.1,  # fee percent (normally 10%)
+            days_remaining=182.5,  # 6 months remaining
+            time_stretch_apy=0.05,  # APY of 5% used to calculate time_stretch
+        ),
+        # From the input, we have the following values:
+        # t_stretch = 22.1868770168519182502689135891
+        # tau = 0.0225358440315970471499308329778
+        # 1 - tau = 0.977464155968402952850069167022
+        # k = c/mu*(mu*z)**(1 - tau) + (2*y + c*z)**(1 - tau)
+        #   = 100000**0.9774641559684029528500691670222 + (2*100000 + 100000*1)**0.9774641559684029528500691670222
+        #   = 302929.51067963685
+        TestResultCalcInGivenOutSuccessByModel(
+            yieldspace=TestResultCalcInGivenOutSuccess(
+                # p = ((2y+cz)/(mu*z))**t
+                #   = 1.0250671833648672
+                # without_fee_or_slippage = 1/p * out = 97.55458141947516
+                without_fee_or_slippage=97.55458141947516,
+                # d_z' = 1/mu * (mu/c*(k - (2*y + c*z - d_y)**(1 - tau)))**(1/(1 - tau)) - z
+                # d_z' = 1/1 * (1/1*(302929.51067963685 - (2*100000 + 100000 - 100)**(1-0.0225358440315970471499308329778)))**(1/(1-0.0225358440315970471499308329778)) - 100000
+                #         = 97.55601990513969
+                without_fee=97.55601990513969,
+                # fee is 10% of discount before slippage = (100-97.55458141947516)*0.1 = 0.24454185805248443
+                fee=0.24454185805248443,
+                # with_fee = d_z' + fee = 97.55601990513969 + 0.24454185805248443 = 97.80056176319217
+                with_fee=97.80056176319218,
+            ),
+            hyperdrive=TestResultCalcInGivenOutSuccess(
+                without_fee_or_slippage=97.58591137152354,
+                without_fee=97.5866001243412,
+                fee=0.24140886284764632,
+                with_fee=97.82800898718885,
```

Comment:
I think we'll ultimately consider this result to be incorrect since this discrepancy arises from our improper use of a normalizing constant for time.

Reply:
ok cool, when we make the change to time, we'll have to fix the tests.